### PR TITLE
moved arrow-typeclasses to arrow-core-data with their respective tests

### PR DESCRIPTION
**Pull request by i-walker**
The purpose was at first to solely move arrow-typeclasses into arrow-core-data. 
But, as there are dependencies between some Datatypes, in terms of codegen. I am trying to move more datatypes into arrow-core-data. According to @raulraja  [List](https://github.com/arrow-kt/arrow/pull/1458#issuecomment-496022046):
`The goal here is to move the least amount of datatypes and resolve the error down below`
```
- [X] AndThen -> arrow-core-data
- [X] Ior -> arrow-core-data
- [X] ListK -> arrow-core-data
- [X] MapK -> arrow-core-data
- [X] NonEmptyList -> arrow-core-data
- [ ] SequenceK -> arrow-core-data
- [X] SetK -> arrow-core-data
- [ ] SortedMap -> arrow-core-data
- [ ] SortedMapK -> arrow-core-data
- [X] Validated -> arrow-core-data
```
#### Error
```
e: error: Arrow's annotations can only be used on Kotlin classes. 
Not valid for error.NonExistentClass
```
Which occurs in `arrow-core-extensions` more specifically, during the kapt Task `kaptKotlin` 
after it create the `NonEmptyListBimonad`

### CODE REVIEW

File: modules/core/arrow-core-data/src/test/kotlin/arrow/core/IorTest.kt [hunk 1]
```diff
@@ -1,33 +1,22 @@
...
 import arrow.core.extensions.hash
 import arrow.core.extensions.semigroup
-import arrow.data.extensions.ior.applicative.applicative
-import arrow.data.extensions.ior.bifunctor.bifunctor
-import arrow.data.extensions.ior.eq.eq
-import arrow.data.extensions.ior.hash.hash
-import arrow.data.extensions.ior.monad.monad
-import arrow.data.extensions.ior.show.show
-import arrow.data.extensions.ior.traverse.traverse
+import arrow.data.*
```
Comment (pakoito) [inline modules/core/arrow-core-data/src/test/kotlin/arrow/core/IorTest.kt]: Ktlint will fail on this

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt [hunk 2]
```diff
@@ -160,11 +150,11 @@ fun <E, A> ValidatedOf<E, A>.findValid(SE: Semigroup<E>, that: () -> Validated<E
 
     { e ->
       that().fold(
-        { ee -> Invalid(SE.run { e.combine(ee) }) },
-        { Valid(it) }
+        { ee -> arrow.core.Invalid(SE.run { e.combine(ee) }) },
```
Comment (raulraja) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/Validated.kt]: these are redundant FQNs since you are in the core package

**Comment by i-walker** [comment]
I'll move Coproduct to arrow-core-data and then remove it in another PR

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt [hunk 3]
```diff
@@ -91,4 +84,6 @@ fun <K, A, B> Map<K, A>.foldLeft(b: Map<K, B>, f: (Map<K, B>, Map.Entry<K, A>) -
 fun <K, A, B> Map<K, A>.foldRight(b: Map<K, B>, f: (Map.Entry<K, A>, Map<K, B>) -> Map<K, B>): Map<K, B> =
   this.entries.reversed().k().foldLeft(b) { x, y: Map.Entry<K, A> -> f(y, x) }
 
-fun <K, V> mapOf(vararg tuple: Tuple2<K, V>): MapK<K, V> = if (tuple.isNotEmpty()) tuple.map { it.a to it.b }.toMap().k() else emptyMap<K, V>().k()
+// TODO; check duplicate in TupleN
```
Comment (pakoito) [inline modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt]: FIXME!

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/listk.kt [hunk 4]
```diff
@@ -182,7 +180,7 @@ interface ListKHash<A> : Hash<ListKOf<A>>, ListKEq<A> {
 
   override fun EQ(): Eq<A> = HA()
 
-  override fun ListKOf<A>.hash(): Int = foldLeft(1) { hash, a ->
+  override fun ListKOf<A>.hash(): Int = fix().foldLeft(1) { hash, a ->
```
Comment (pakoito) [inline modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/listk.kt]: 👍

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/nonemptylist.kt [hunk 5]
```diff
@@ -105,7 +104,7 @@ interface NonEmptyListComonad : Comonad<ForNonEmptyList> {
     fix().map(f)
 }
 
-@extension
+/*@extension
```
Comment (pakoito) [inline modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/nonemptylist.kt]: FIXME!

File: modules/core/arrow-extras-data/build.gradle [hunk 6]
```diff
@@ -17,4 +17,4 @@ dependencies {
     testCompile project(':arrow-effects-extensions')
 }
 
-apply from: rootProject.file('gradle/gradle-mvn-push.gradle')
```
Comment (pakoito) [inline modules/core/arrow-extras-data/build.gradle]: Missing: additional endline

File: modules/core/arrow-extras-data/build.gradle [hunk 7]
```diff
@@ -3,7 +3,7 @@ apply plugin: 'kotlin-kapt'
 dependencies {
     compile "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlinVersion"
     compile project(':arrow-annotations')
-    compile project(':arrow-typeclasses')
+    compile project(':arrow-core-data') // has to be deleted in another PR
```
Comment (pakoito) [inline modules/core/arrow-extras-data/build.gradle]: TODO + a link to the github issue!

File: modules/core/arrow-extras-data/src/main/kotlin/arrow/data/SortedMapK.kt [hunk 8]
```diff
@@ -49,7 +50,7 @@ data class SortedMapK<A : Comparable<A>, B>(private val map: SortedMap<A, B>) :
 
   fun <G, C> traverse(GA: Applicative<G>, f: (B) -> Kind<G, C>): Kind<G, SortedMapK<A, C>> = GA.run {
     map.iterator().iterateRight(Eval.always { just(sortedMapOf<A, C>().k()) }) { kv, lbuf ->
-      f(kv.value).map2Eval(lbuf) { (mapOf(kv.key to it.a).k() + it.b).toSortedMap().k() }
+      f(kv.value).map2Eval(lbuf) { (mapOf(kv.key to it.a).mk() + it.b).toSortedMap().k() }
```
Comment (pakoito) [inline modules/core/arrow-extras-data/src/main/kotlin/arrow/data/SortedMapK.kt]: mk?

File: modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt [hunk 9]
```diff
@@ -76,7 +77,7 @@ class StateT<F, S, A>(
      * @param fa the value to liftF.
      */
     fun <F, S, A> liftF(AF: Applicative<F>, fa: Kind<F, A>): StateT<F, S, A> = AF.run {
-      StateT(just({ s -> fa.map { a -> Tuple2(s, a) } }))
+      StateT(just { s -> fa.map { a -> Tuple2(s, a) } })
```
Comment (pakoito) [inline modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt]: ...this looks wrong. The lambda is a parameter.

File: modules/core/arrow-extras-data/src/test/kotlin/arrow/data/SequenceKTest.kt [hunk 10]
```diff
@@ -21,7 +21,7 @@ import arrow.test.laws.MonoidLaws
 import arrow.test.laws.MonoidalLaws
 import arrow.test.laws.ShowLaws
 import arrow.test.laws.TraverseLaws
-import arrow.test.laws.fix
```
Comment (pakoito) [inline modules/core/arrow-extras-data/src/test/kotlin/arrow/data/SequenceKTest.kt]: Remove or fix

File: modules/core/arrow-extras-data/src/main/kotlin/arrow/data/SortedMapK.kt (hunk 8 above)
Comment (i-walker) [reply modules/core/arrow-extras-data/src/main/kotlin/arrow/data/SortedMapK.kt]: it was necessary the file has multiple k() function from different imports

File: modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt (hunk 9 above)
Comment (i-walker) [reply modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt]: that was what kotlin suggested.

File: modules/optics/arrow-optics/src/test/kotlin/arrow/optics/std/EitherTest.kt [hunk 11]
```diff
@@ -26,14 +26,14 @@ class EitherTest : UnitSpec() {
         when (this) {
           is Invalid -> {
             when (b) {
-              is Invalid -> Invalid((e + b.e))
+              is Invalid -> arrow.core.Invalid((e + b.e))
```
Comment (pakoito) [inline modules/optics/arrow-optics/src/test/kotlin/arrow/optics/std/EitherTest.kt]: Fix these too!

File: modules/optics/arrow-optics/src/test/kotlin/arrow/optics/std/TryTest.kt [hunk 12]
```diff
@@ -71,14 +71,14 @@ class TryTest : UnitSpec() {
           when (this) {
             is Invalid -> {
               when (b) {
-                is Invalid -> Invalid(e)
+                is Invalid -> arrow.core.Invalid(e)
```
Comment (pakoito) [inline modules/optics/arrow-optics/src/test/kotlin/arrow/optics/std/TryTest.kt]: Unnecessary qualifiers

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt (hunk 3 above)
Comment (i-walker) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt]: The problem is that there are 2 identical functions. Both of them lived in arrow-core-data and arrow-extras-data. 
I commented out one. We need to decide which should live.
Screenshot 2019-06-15 at 01 44 30

File: modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/nonemptylist.kt (hunk 5 above)
Comment (i-walker) [reply modules/core/arrow-core-extensions/src/main/kotlin/arrow/core/extensions/nonemptylist.kt]: upps

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt (hunk 3 above)
Comment (pakoito) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt]: This one lives I'd say. Where's the arrow-extras-data one?

File: modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt (hunk 9 above)
Comment (pakoito) [reply modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt]: I believe that.

Comment (i-walker) [reply modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt]: So, should I revert the change?

File: modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt (hunk 3 above)
Comment (i-walker) [reply modules/core/arrow-core-data/src/main/kotlin/arrow/core/MapK.kt]: Here https://github.com/arrow-kt/arrow/blob/d1b7865c7383c58cee04447a7487190635505c9a/modules/core/arrow-core-data/src/main/kotlin/arrow/core/TupleN.kt#L152
But now MapK is in arrow-core-data, hence the duplication

**Comment by pakoito** [comment]
Great job, it took a lot of effort to get it done and it seems mostly correct. Let's merge :D

File: modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt (hunk 9 above)
Comment (raulraja) [reply modules/core/arrow-extras-data/src/main/kotlin/arrow/data/StateT.kt]: When lambdas are single parameter you can omit the braces and just inline the function. StateT wraps a function.

**Pull request merged**